### DISCUSSION TITLE
How do people handle fetching only the required set of fields for each use?

### DISCUSSION THREAD

**Discussion by Zeblote** [diesel-rs/diesel#2813] [Q&A]
I'm wondering how people handle only requesting the fields that you actually need for each individual query. For example, if your user table has 30 columns, and you just want their id and name, and one specific settings value, clearly it doesn't make sense to fetch all of them. I can only imagine few situations where I'd actually want _all_ fields of a given table. Most queries require different, potentially overlapping sets of fields, and while you _could_ fetch all of them, that would only result in worse db performance.

In js world, using sequelize, we made frequent use of `attributes: [ 'a', 'b', 'f' ]` to fetch only the necessary ones. However, this is completely unsafe since you can (and I have) forget to request a field and then use it later anyway, causing lots of trouble at runtime...

It seems like the intended solution using diesel is to create a new struct for each set of fields you want. However, that quickly becomes unmaintainable since you have to annotate every single one with all the database types and you also lose the ability to make shared utility functions that can take a user and do some things with it.

Has anyone found a usable solution for the following:
- Be able to request a specifc subset of fields to be returned for every individual query
- Be able to pass the result of that to shared utilitity functions that take objects from this table
- Get a compile error if the utility function tries to access a field that was not included in the query
- Not have to implement a ridiculous amount of boilerplate code to achieve this

**Comment by weiznich** [comment]
First of all I believe that is a fundamental "issue". You have to choose between two options here, flexibility on the cost of static checking or hard type safety on the cost of writing a bit more code. You cannot have both, as the only way to communicate that a certain field was load or not at compile time is to actually have different types those both cases. In the end it boils down to that each type in rust needs to have exactly one concrete representation, which implies that if your query returns a different set of fields you have to choose between having a overhead there (for having unused field), or having different types. Giving concrete proposals on how to solve your problem is hard as it misses a real example.

Other than that let me clarify a few other points:

* First of all structs implementing `Queryable` are **not** designed to be a single coherent model that maps 1:1 to table. They are meant to be used 1:1 with queries, where a query can easily contain more than one table. It's mainly designed about the use case of a json endpoint somewhere that returns exactly one result. In that case the struct must already exist for having `#[derive(Serialize)]` on it.
* You don't need to have separate structs for each query, as tuples also implement `Queryable`
* The `Queryable` derive does not depend on any database related information. You don't need anything else than `#[derive(Queryable)]` on a struct.

There may be a way to implement something roughly like this in at type level, but be warned that this is just a rough sketch, that is in no way supported by diesel. Implementing this requires advanced knowledge of traits and generics in rust. Expect where clause with >> 10 lines everywhere. 
It may be possible to create clever type wrappers that allow you to have a "single" generic container here that allows to access inner opaque fields via a key checked at compile time, but even in that case you get different types for all the different variants. Think of that variant as having a `User<Some, Generic, Inner, Types, …>`, where the generic arguments are different for each case. [frunk](https://docs.rs/frunk/0.4.0/frunk/labelled/trait.LabelledGeneric.html) implements some basic helpers that could be useful here, but again be warned that this is likely not the solution you want to choose, as this requires a really large amount of really complex generic code.

**Reply by Yoric** [reply]
> First of all structs implementing Queryable are not designed to be a single coherent model that maps 1:1 to table. They are meant to be used 1:1 with queries, where a query can easily contain more than one table. 

FWIW, I believe that this is one of the key concept that I had difficulties grasping with Diesel.

I imagine that there are ways to make it clearer within the documentation and/or error messages. Not sure how :)

**Reply by weiznich** [reply]
@Yoric Feel free to suggest concrete improvements for the documentation there. Its always hard for us maintainers to know where people would look for such information.
